libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,379 @@
+			if (_op.lengthParameter)
+				ourOp.length = identifierNameIfSSA(_functionCall.arguments.at(*_op.lengthParameter));
+			if (_op.lengthConstant)
+				switch (*_op.lengthConstant)
+				{
+				case 1: ourOp.length = YulString(one); break;
+				case 32: ourOp.length = YulString(thirtyTwo); break;
+				default: yulAssert(false);
+				}
+			return ourOp;
+		}
+	);
+}
+
+void UnusedStoreEliminator::applyOperation(UnusedStoreEliminator::Operation const& _operation)
+{
+	for (auto& [statement, state]: m_stores[YulString{}])
+		if (state == State::Undecided)
+		{
+			Operation const& storeOperation = m_storeOperations.at(statement);
+			if (_operation.effect == Effect::Read && !knownUnrelated(storeOperation, _operation))
+				state = State::Used;
+			else if (_operation.effect == Effect::Write && knownCovered(storeOperation, _operation))
+				state = State::Unused;
+		}
+}
+
+bool UnusedStoreEliminator::knownUnrelated(
+	UnusedStoreEliminator::Operation const& _op1,
+	UnusedStoreEliminator::Operation const& _op2
+) const
+{
+	KnowledgeBase knowledge(m_dialect, m_ssaValues);
+
+	if (_op1.location != _op2.location)
+		return true;
+	if (_op1.location == Location::Storage)
+	{
+		if (_op1.start && _op2.start)
+		{
+			yulAssert(
+				_op1.length &&
+				_op2.length &&
+				knowledge.valueIfKnownConstant(*_op1.length) == 1 &&
+				knowledge.valueIfKnownConstant(*_op2.length) == 1
+			);
+			return knowledge.knownToBeDifferent(*_op1.start, *_op2.start);
+		}
+	}
+	else
+	{
+		yulAssert(_op1.location == Location::Memory, "");
+		if (
+			(_op1.length && knowledge.knownToBeZero(*_op1.length)) ||
+			(_op2.length && knowledge.knownToBeZero(*_op2.length))
+		)
+			return true;
+
+		if (_op1.start && _op1.length && _op2.start)
+		{
+			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
+			optional<u256> start1 = knowledge.valueIfKnownConstant(*_op1.start);
+			optional<u256> start2 = knowledge.valueIfKnownConstant(*_op2.start);
+			if (
+				(length1 && start1 && start2) &&
+				*start1 + *length1 >= *start1 && // no overflow
+				*start1 + *length1 <= *start2
+			)
+				return true;
+		}
+		if (_op2.start && _op2.length && _op1.start)
+		{
+			optional<u256> length2 = knowledge.valueIfKnownConstant(*_op2.length);
+			optional<u256> start2 = knowledge.valueIfKnownConstant(*_op2.start);
+			optional<u256> start1 = knowledge.valueIfKnownConstant(*_op1.start);
+			if (
+				(length2 && start2 && start1) &&
+				*start2 + *length2 >= *start2 && // no overflow
+				*start2 + *length2 <= *start1
+			)
+				return true;
+		}
+
+		if (_op1.start && _op1.length && _op2.start && _op2.length)
+		{
+			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
+			optional<u256> length2 = knowledge.valueIfKnownConstant(*_op2.length);
+			if (
+				(length1 && *length1 <= 32) &&
+				(length2 && *length2 <= 32) &&
+				knowledge.knownToBeDifferentByAtLeast32(*_op1.start, *_op2.start)
```

Comment:
Hm... this could be
```knownToBeDifferentByAtLeast(*_op.start, *_op2.start, max(*lenth1, *length2))```
thinking of two ``mstore8``s...

And there's nothing really that restricts this case to only ``<= 32``, is there?

Reply:
Probably ``<= 2^255`` is the only boundary, if any?

Reply:
That is true, but I'm not sure it is worth it delaying the PR :)